compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde.hs

```diff
@@ -0,0 +1,500 @@
+      putMVar (initParamsVar miState) _params
+      sendClient miState $ LSP.FromServerRsp _method $ LSP.ResponseMessage "2.0" (Just _id) (Right initializeResult)
+    LSP.FromClientMess LSP.SWindowWorkDoneProgressCancel notif -> do
+      (newNotif, mHome) <- removeWorkDoneProgressCancelTokenPrefix (progessTokenPrefixesVar miState) notif
+      let newMsg = LSP.FromClientMess LSP.SWindowWorkDoneProgressCancel newNotif
+      case mHome of
+        Nothing -> void $ sendAllSubIDEs miState newMsg
+        Just home -> void $ sendSubIDEByPath miState home newMsg
+
+    -- Special handing for STextDocumentDefinition to ask multiple IDEs (the W approach)
+    -- When a getDefinition is requested, we cast this request into a tryGetDefinition
+    -- This is a method that will take the same logic path as getDefinition, but will also return an
+    -- identifier in the cases where it knows the identifier wasn't defined in the package that referenced it
+    -- When we receive this name, we lookup against the multi-package.yaml for a package that matches where the identifier
+    -- came from. If we find one, we ask (and create if needed) the SubIDE that contains the identifier where its defined.
+    -- (this is via the getDefinitionByName message)
+    -- We also send the backup known incorrect location from the tryGetDefinition, such that if the subIDE containing the identifier
+    -- can't find the definition, it'll fall back to the known incorrect location.
+    -- Once we have this, we return it as a response to the original STextDocumentDefinition request.
+    LSP.FromClientMess LSP.STextDocumentDefinition req@LSP.RequestMessage {_id, _method, _params} -> do
+      let path = filePathFromParamsWithTextDocument req
+          lspId = castLspId _id
+          method = LSP.SCustomMethod "daml/tryGetDefinition"
+      putReqMethodSingleFromClient (fromClientMethodTrackerVar miState) lspId method
+      sendSubIDEByPath miState path $ LSP.FromClientMess method $ LSP.ReqMess $
+        LSP.RequestMessage "2.0" lspId method $ Aeson.toJSON $
+          TryGetDefinitionParams (_params ^. LSP.textDocument) (_params ^. LSP.position)
+
+    LSP.FromClientMess meth params ->
+      case getMessageForwardingBehaviour meth params of
+        ForwardRequest mess (Single path) -> do
+          debugPrint $ "single req on method " <> show meth <> " over path " <> path
+          let LSP.RequestMessage {_id, _method} = mess
+          putReqMethodSingleFromClient (fromClientMethodTrackerVar miState) _id _method
+          sendSubIDEByPath miState path (castFromClientMessage msg)
+
+        ForwardRequest mess (AllRequest combine) -> do
+          debugPrint $ "all req on method " <> show meth
+          let LSP.RequestMessage {_id, _method} = mess
+          ides <- sendAllSubIDEs miState (castFromClientMessage msg)
+          putReqMethodAll (fromClientMethodTrackerVar miState) _id _method ides combine
```

Comment:
as discussed, later we should think about using `with`-style functions on the `TVar` to avoid the scenario where a response comes before `putReqMethodAll`

Reply:
Part of the refactor ticket :)